R/dist_spec.R

```diff
@@ -235,53 +259,57 @@ mean.dist_spec <- function(x, ..., ignore_uncertainty = FALSE) {
 #' @return A vector of standard deviations.
 #' @importFrom utils head
 #' @keywords internal
+#' @export
 #' @examples
 #' \dontrun{
 #' # A fixed lognormal distribution with sd 5 and sd 1.
 #' dist1 <- LogNormal(mean = 5, sd = 1, max = 20)
-#' sd_dist(dist1)
+#' sd(dist1)
 #'
 #' # A gamma distribution with mean 3 and sd 2
 #' dist2 <- Gamma(mean = 3, sd = 2)
-#' sd_dist(dist2)
+#' sd(dist2)
 #'
 #' # The sd of the sum of two distributions
-#' sd_dist(dist1 + dist2)
+#' sd(dist1 + dist2)
 #' }
-sd_dist <- function(x) {
-  ret <- vapply(x, function(y) {
-    if (is.numeric(y)) {
-      return(0)
+sd.dist_spec <- function(x, ...) {
+  if (x$distribution == "nonparametric") {
+    ## nonparametric
+    mean_pmf <- sum((seq_along(x$pmf) - 1) * x$pmf)
+    return(sum((seq_along(x$pmf) - 1)**2 * x$pmf) - mean_pmf^2)
+  } else {
+    ## parametric
+    if (!all(vapply(x$parameters, is.numeric, logical(1)))) {
+      return(NA_real_)
```

Comment:
Should this warn at least?

Reply:
I think we use this a fair bit internally (the post-processing the NAs) so this would produce a lot of warnings. I've added a note to the documentation instead in 07eb4fb17d29524271ec65cc08ada262d11c6089